plugins/modules/proxmox_vm_info.py

```diff
@@ -154,42 +168,51 @@ def get_vms_from_cluster_resources(self):
                 msg="Failed to retrieve VMs information from cluster resources: %s" % e
             )
 
-    def get_vms_from_nodes(self, vms_unfiltered, type, vmid=None, name=None, node=None):
-        vms = []
+    def get_vms_from_nodes(self, vms_unfiltered, type, vmid=None, name=None, node=None, config=None):
+        filtered_vms = []
         for vm in vms_unfiltered:
             if (
-                type != vm["type"]
-                or (node and vm["node"] != node)
-                or (vmid and int(vm["vmid"]) != vmid)
-                or (name is not None and vm["name"] != name)
+                    type != vm["type"]
+                    or (node and vm["node"] != node)
+                    or (vmid and int(vm["vmid"]) != vmid)
+                    or (name is not None and vm["name"] != name)
             ):
                 continue
-            vms.append(vm)
-        nodes = frozenset([vm["node"] for vm in vms])
-        for node in nodes:
+            filtered_vms.append(vm)
+        nodes = frozenset([filtered_vm["node"] for filtered_vm in filtered_vms])
+        for this_node in nodes:
             if type == "qemu":
-                vms_from_nodes = self.proxmox_api.nodes(node).qemu().get()
+                vms_from_this_node = self.proxmox_api.nodes(this_node).qemu().get()
             else:
-                vms_from_nodes = self.proxmox_api.nodes(node).lxc().get()
-            for vmn in vms_from_nodes:
-                for vm in vms:
-                    if int(vm["vmid"]) == int(vmn["vmid"]):
-                        vm.update(vmn)
-                        vm["vmid"] = int(vm["vmid"])
-                        vm["template"] = proxmox_to_ansible_bool(vm["template"])
+                vms_from_this_node = self.proxmox_api.nodes(this_node).lxc().get()
+            for detected_vm in vms_from_this_node:
+                for desired_vm in filtered_vms:
+                    if int(desired_vm["vmid"]) == int(detected_vm["vmid"]):
```

Comment:
How about creating a map `desired_vm` mapping the vmid to the `desired_vm` object? Then the above could be changed to
```py
            for detected_vm in vms_from_this_node:
                desired_vm = desired_vm.get(int(detected_vm["vmid"]))
                if desired_vm:
```
Especially if `filtered_vms` is large, this would improve performance a lot.

Reply:
I'm afraid it will not work - both `vms_from_this_node` and `filtered_vms` are plain lists of similar dictionaries (which sometimes have equal `vmid` values).

Reply:
But isn't your code (which assumes that you can uniquely identify an entry by its `vmid`) also wrong?

Reply:
I know that triple nested loops are evil but I wanted initially make the least injection to someone else's code. :)

Nested loop used to find "objects" of the same VM in both lists. Condition `int(desired_vm["vmid"]) == int(detected_vm["vmid"])` is `true` when VM is present in both lists (list of VMs that user want to retrieve - `filtered_vms` - and list of VMs that script found on particular node - `vms_from_this_node`).
My part of code is called only in the case when the script have found some VM which user wants to know about and this VM is placed on the node which is being processed at this iteration of the outer loop.
After successful processing of found (in both lists) VM the inner loop is breaking.

Reply:
But shouldn't you also check `desired_vm['node']` in this `if`?

Reply:
I think it is unnecessary, because VMID is unique inside a Proxmox cluster. There will be no other machine with the same VMID - no need to check node where it is placed.

Reply:
But if vmid is unique, why did my original suggestion of using a dictionary based on vmid not work? I'm confused now :)

Reply:
>  use that dict and not the list.

Made it. And fixed one hidden bug - I am not using lxc containers and forgot to add config getter to my part of code.